arch/powerpc/platforms/pseries/eeh_pseries.c

```diff
@@ -793,6 +793,69 @@ static int pseries_eeh_restore_config(struct pci_dn *pdn)
 	return 0;
 }
 
+static int pseries_call_allow_unfreeze(struct eeh_dev *edev)
+{
+	struct eeh_pe *pe;
+	struct pci_dn *pdn;
+	int cur_vfs, rc, config_addr, vf_index;
+	u16 *vf_pe_array;
+	int ibm_allow_unfreeze = rtas_token("ibm,open-sriov-allow-unfreeze");
+
+	if (ibm_allow_unfreeze == RTAS_UNKNOWN_SERVICE)
+		return 0;
+
+	pe = eeh_dev_to_pe(edev);
+	cur_vfs = pci_num_vf(edev->pdev);
+	rc = 0;
+
+	if (cur_vfs) {
+
+		/* For each of its vf
+		 * call allow unfreeze
+		 */
+
+		config_addr = pe->config_addr;
+		vf_pe_array = kzalloc(RTAS_DATA_BUF_SIZE, GFP_KERNEL);
+		if (!vf_pe_array)
+			return -EINVAL;
+
+		pdn = eeh_dev_to_pdn(edev);
+
+		for (vf_index = 0; vf_index < cur_vfs; vf_index++)
+			vf_pe_array[vf_index] =
+				be16_to_cpu(pdn->pe_num_map[vf_index]);
+
+		spin_lock(&rtas_data_buf_lock);
+		memcpy(rtas_data_buf, vf_pe_array, cur_vfs*sizeof(u16));
```

Comment:
You should be using RTAS_DATA_BUF_SIZE since vf_pe_array is initialized to that size and so is rtas_data_buf otherwise you might get overflow or underflow issues.

Reply:
will change.